appcelerator/titanium_mobile
# [TIMOB-16878] Android: Refactored ScrollableView

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-5996
- https://jira.appcelerator.org/browse/TIMOB-15780
- https://jira.appcelerator.org/browse/TIMOB-16878
- https://jira.appcelerator.org/browse/TIMOB-25539
- https://jira.appcelerator.org/browse/TIMOB-25625
- https://jira.appcelerator.org/browse/TIMOB-25626
- https://jira.appcelerator.org/browse/TIMOB-25634
- https://jira.appcelerator.org/browse/TIMOB-25636

**Summary:**
- [TIMOB-5996](https://jira.appcelerator.org/browse/TIMOB-5996) Modified `ScrollableView.removeView()` to support integer indexes like iOS.
- [TIMOB-15780](https://jira.appcelerator.org/browse/TIMOB-15780) Fixed `ScrollableView.setCurrentPage()` to not fire a "scrollend" event.
  * Event should only be fired when dragged or for methods that trigger scrolling animations.
  * Added missing "dragstart" event to be paired with existing "dragend" event. Now matches docs.
- [TIMOB-16878](https://jira.appcelerator.org/browse/TIMOB-16878) Fixed bug where ScrollableView ignores top-most child view's layout properties.
  * Prevented properties width, height, top, bottom, left, and right from working.
  * Fixed by wrapping view with a `TiCompositeLayout`, which is needed to handle these layout properties.
- [TIMOB-25539](https://jira.appcelerator.org/browse/TIMOB-25539) Fixed crash caused by adding a border to the ScrollableView.
  * Bug was introduced in Titanium 6.1.0.
- [TIMOB-25625](https://jira.appcelerator.org/browse/TIMOB-25625) Modified left/right page arrows to be density scaled.
  * Was hard-code to 80 pixels for all screens.
  * Now uses base mdpi (1x) resolution of 24 pixels and scales based on DPI from there.
- [TIMOB-25626](https://jira.appcelerator.org/browse/TIMOB-25626) Fixed ScrollableView to support `Ti.UI.SIZE` for width.
  * Android's `ViewPager` Java class does not natively support `WRAP_CONTENT`.
  * Refactored onMeasure() code to re-size width/height of container based on largest child view in this case.
- [TIMOB-25634](https://jira.appcelerator.org/browse/TIMOB-25634) Modified ScrollableView so that `cacheSize` property can be set dynamically like iOS.
  * Added methods `setCacheSize()` and `getCacheSize()`.
  * Can no longer be set less than 3, matching iOS' min cache size.
  * Now better converts iOS style cache size value to Android offscreen page count value.
- [TIMOB-25636](https://jira.appcelerator.org/browse/TIMOB-25636) Fixed bug where calling ScrollableView setters cause a crash if parent window is closed.
  * Bug was introduced as of Titanium 7.0.0.
- Removed API Level 9 guards from code.
  * Outdated code. Min API Level is currently 16.

---
**Test 1:**
This test ensures border applied to ScrollableView does not cause a crash.
1. Build and run the code attached to [TIMOB-25539](https://jira.appcelerator.org/browse/TIMOB-25539) on Android.
2. Verify that the app does not crash on startup.
3. Verify that a blue border is displayed around the ScrollableView.

---
**Test 2:**
Test that left/right page arrows are now density scaled. Note that the overlaid arrows used to be too tiny on high DPI devices.
1. Build and run the code attached to [TIMOB-25625](https://jira.appcelerator.org/browse/TIMOB-25625) on an "hdpi" and "xxhdpi" Android devices.
2. Scroll to the right in the ScrollableView on both devices.
3. Verify that the overlaid left/right page arrows are about the same size. (Arrows should no longer be tiny on a xxxhdpi device.)

---
**Test 3:**
Tests ScrollableView Ti.UI.SIZE support for [TIMOB-25626](https://jira.appcelerator.org/browse/TIMOB-25626). Also tests layout fix for [TIMOB-16878](https://jira.appcelerator.org/browse/TIMOB-16878).
1. Build and run the code attached to [TIMOB-25626](https://jira.appcelerator.org/browse/TIMOB-25626) on Android.
2. Note that the light gray container is the ScrollableView and the dark gray area is the Window.
3. Verify that view looks like [Android-good1.png](https://jira.appcelerator.org/secure/attachment/65127/Android-good1.png) on startup. Light gray container should not fill the window. (That's the old bad behavior.)
4. Scroll the light gray view to the right (by swiping left).
5. Verify view looks like [Android-good2.png](https://jira.appcelerator.org/secure/attachment/65123/Android-good2.png). Container should grow a bit bigger.
6. Scroll the light gray view to the right (by swiping left).
7. Verify view looks like [Android-good3.png](https://jira.appcelerator.org/secure/attachment/65126/Android-good3.png). container should grow a bit bigger.
8. Scroll the light gray view to the right (by swiping left).
9. Verify view looks like [Android-good4.png](https://jira.appcelerator.org/secure/attachment/65122/Android-good4.png). container should grow a bit bigger.
10. Scroll the light gray view to the right (by swiping left).
11. Purple view should now completely fill light gray container like [Android-good5.png](https://jira.appcelerator.org/secure/attachment/65125/Android-good5.png).

---
**Test 4:**
Test dynamically changing `cacheSize` property.
1. Build and run file `ScrollableViewCacheSizeTest.js` attached to [TIMOB-25634](https://jira.appcelerator.org/browse/TIMOB-25634) on Android.
2. Verify that the following gets logged...
```
[INFO] :   @@@ Initialized cacheSize: 5
[INFO] :   @@@ Property assigned cacheSize: 4
[INFO] :   @@@ Method assigned cacheSize: 3
```

---
**Test 5:**
1. Build and run file `ScrollableViewAddRemoveTest.js` attached to [TIMOB-15780](https://jira.appcelerator.org/browse/TIMOB-15780) on Android.
2. Note that this app automatically cycles through the pages programmatically via a timer.
2. Watch the Android log.
3. Verify that `@@@ [scrollend]` does not get logged when it programmatically changes pages. This is the fix for [TIMOB-15780](https://jira.appcelerator.org/browse/TIMOB-15780).
4. Scroll to another page with your finger. (Swipe left or right.)
5. Verify that a `[dragstart]`, `[scroll]`, `[dragend]`, and `[scrollend]` get logged when you scroll via the touchscreen yourself.
6. Tap the "Remove" button in the bottom-right corner.
7. An alert dialog is displayed. Tap the "Current" button to remove current page. This removes page by index and tests [TIMOB-5996](https://jira.appcelerator.org/browse/TIMOB-5996).
8. Repeat steps 6 and 7 until all 4 pages have been removed.
